onnxruntime/core/providers/cuda/tensor/transpose.cc

```diff
@@ -92,20 +92,76 @@ Status Transpose::DoTranspose(const Transpose& kernel,
   const std::vector<int64_t>& output_dims = output.Shape().GetDims();
 
   auto rank = static_cast<int32_t>(input_dims.size());
-  TensorPitches original_input_strides(input_dims);
-  TensorPitches original_output_strides(output_dims);
 
-  TArray<int64_t> input_strides(rank);
-  for (auto i = 0; i < rank; i++) {
-    input_strides[i] = original_input_strides[permutations[i]];
-  }
-  TArray<fast_divmod> output_strides(rank);
-  for (auto i = 0; i < rank; i++) {
-    output_strides[i] = fast_divmod(gsl::narrow_cast<int>(original_output_strides[i]));
+  // flatten the adjacent dimensions which are contiguous
+  // for example: permutations[0, 2, 3, 1] -> [0, 2, 1], permutations[0, 3, 1, 2] -> [0, 2, 1]
+  auto new_rank = rank;
+  std::vector<size_t> new_permutations(permutations);
+  std::vector<int64_t> new_input_dims(input_dims);
+  std::vector<int64_t> new_output_dims(output_dims);
+  for (auto i = rank - 1; i > 0; i--) {
+    auto curr = permutations[i];
+    auto prev = permutations[i - 1];
+    if (prev + 1 == curr) {
+      // all dims bigger than curr need to be reduced by 1 due to the merging.
+      for (auto j = 0; j < new_rank; j++) {
```

Comment:
>j [](start = 16, length = 1)

the variable `j` is used in different contexts a few times, i think the various j's could benefit from more meaningful names

Reply:
j is just used as an index to access different arrays.

Reply:
all js are basically index of arrays. No different meaning here.